test/pipeline/pipeline-test.cpp

```diff
@@ -0,0 +1,237 @@
+    // for more information
+
+    // The first stage goes from an integer to a string
+    std::unique_ptr<MockStage<int, std::string>> stage1(new MockStage<int, std::string>());
+    EXPECT_CALL(*stage1, Run(integers[test_set]))
+        .WillOnce(testing::Return(strings[test_set]));
+    // This line (.WillOnce(...);), we have here to make sure that this function is only called once
+
+    // The second stage takes a string and returns
+    // a char
+    std::unique_ptr<MockStage<std::string, char>> stage2(new MockStage<std::string, char>());
+    EXPECT_CALL(*stage2, Run(strings[test_set]))
+        .WillOnce(testing::Return(characters[test_set]));
+
+    // Now, we construct the pipeline and run it
+    char result = pipeline.AddStage(*stage1)
+                          .Complete(*stage2)
+                          .Run(integers[test_set]);
+    // And we verify the result
+    ASSERT_EQ(characters[test_set], result);
+}
+// DOUBLE > FLOAT > STRING
+TEST_F(PipelineTest, DoubleToFloatToStringPipeline) {
+    INIT_DOUBLE_TO_STRING_PIPELINE(pipeline, stages);
+
+    int test_set = 0;
+
+    std::unique_ptr<MockStage<double, float>> stage1(new MockStage<double, float>());
+    EXPECT_CALL(*stage1, Run(doubles[test_set]))
+        .WillOnce(testing::Return(floats[test_set]));
+
+    std::unique_ptr<MockStage<float, std::string>> stage2(new MockStage<float, std::string>());
+    EXPECT_CALL(*stage2, Run(floats[test_set]))
+        .WillOnce(testing::Return(strings[test_set]));
+
+    std::string result = pipeline.AddStage(*stage1)
+                          .Complete(*stage2)
+                          .Run(doubles[test_set]);
+    ASSERT_EQ(std::string(strings[test_set]), result);
+}
+
+// CHAR > INT > Double
+TEST_F(PipelineTest, CharToIntToDoublePipeline) {
+    INIT_CHAR_TO_DOUBLE_PIPELINE(pipeline, stages);
+
+    int test_set = 0;
+
+    std::unique_ptr<MockStage<char, int>> stage1(new MockStage<char, int>());
+    EXPECT_CALL(*stage1, Run(characters[test_set]))
+        .WillOnce(testing::Return(integers[test_set]));
+
+    std::unique_ptr<MockStage<int, double>> stage2(new MockStage<int, double>());
+    EXPECT_CALL(*stage2, Run(integers[test_set]))
+        .WillOnce(testing::Return(doubles[test_set]));
+
+    double result = pipeline.AddStage(*stage1)
+                          .Complete(*stage2)
+                          .Run(characters[test_set]);
+    ASSERT_EQ(doubles[test_set], result);
+}
+
+// one stage pipeline
+TEST_F(PipelineTest, SingleStagePipeline) {
+    INIT_INT_TO_CHAR_PIPELINE(pipeline, stages);
+
+    int test_set = 0;
+
+    std::unique_ptr<MockStage<int, char>> stage1(new MockStage<int, char>());
+    EXPECT_CALL(*stage1, Run(integers[test_set]))
+        .WillOnce(testing::Return(characters[test_set]));
+
+    char result = pipeline.Complete(*stage1)
+                          .Run(integers[test_set]);
+
+    ASSERT_EQ(characters[test_set], result);
+}
+
+// empty pipeline > expecting failure
```

Comment:
Make your description more descriptive. It should explain that the failure is connected with the `Run` function. Same with the name of the test.